Enforce chunks ordering when writing index.

pstibrany opened the pull request:
This PR enforces ordering of chunk references when adding series to the index. This is done to avoid direct users of `chunks.Writer` and `index.Writer` to write chunks unordered. Prometheus itself already writes chunks ordered.

bwplotka reviewed (approved):
Nice, I think this is great but we talked with @codesome and we have 2 suggestions:

* Can we update format.md for this invariant?
* Can we also ensure that chunks are put together for the series. Wonder if it's efficient to check it, but maybe we can have at least the test AND docs :thinking: 

Thanks!

roidelapluie commented:
Since https://github.com/cortexproject/cortex/pull/3371 is merged, I think that this is no longer needed, but that we could still have the documentation update. Is that correct? Thanks!

pstibrany commented:
I think this PR would still be nice to get in, so that others don’t make the same mistake. I have forgotten about it, but will try to finish it (address Bartek’s suggestions) soon. Thanks for patience.

roidelapluie commented:
Awesome 😎

pstibrany commented:
> Nice, I think this is great but we talked with @codesome and we have 2 suggestions:
> 
> * Can we update format.md for this invariant?

Added.

> * Can we also ensure that chunks are put together for the series. Wonder if it's efficient to check it, but maybe we can have at least the test AND docs 🤔

Check for increasing chunk ref number already guarantees this property.

I've updated the PR with checks for ordering of `minT` and `maxT` for chunks **in the same series**:
- for chunk, `c.minT` must be less or equal to `c.maxT`
- for two subsequent chunks, `c_i.maxT` < `c_(i+1).minT` must hold

These are already implicit in encoding of minT/maxT in the index.

roidelapluie commented:
Tests are failing.

pstibrany commented:
I think this is still relevant, I will update the PR to resolve conflicts.

pstibrany commented:
> * for two subsequent chunks, `c_i.maxT` < `c_(i+1).minT` must hold

Funnily enough, this check would help us to detect https://github.com/prometheus/prometheus/issues/12873 earlier :)

pstibrany commented:
Any chance for getting this merged, so that I don't need to rebase again? :)

beorn7 commented:
:wave: from the bug scrub again. Thank you very much @pstibrany. This is almost good to go. @bboreham will have a last look.

bboreham reviewed (approved):
lgtm
